zplugin-ros1/src/ros_to_zenoh_bridge/zenoh_client.rs

```diff
@@ -0,0 +1,148 @@
+//
+// Copyright (c) 2022 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::sync::Arc;
+
+use log::{debug, info};
+
+use zenoh::prelude::r#async::*;
+use zenoh::Session;
+pub use zenoh_core::zresult::ZResult;
+
+pub struct ZenohClient {
+    session: Arc<Session>,
+}
+
+impl ZenohClient {
+    // PUBLIC
+    pub fn new(session: Arc<Session>) -> ZenohClient {
+        ZenohClient { session }
+    }
+
+    pub async fn subscribe<C>(
+        &self,
+        key_expr: &str,
+        callback: C,
+    ) -> ZResult<zenoh::subscriber::Subscriber<'static, ()>>
+    where
+        C: Fn(Sample) + Send + Sync + 'static,
+    {
+        debug!("Creating Subscriber on {}", key_expr);
+
+        match self.make_keyexpr(key_expr).await {
+            Ok(opt_keyexpr) => {
+                self.session
+                    .declare_subscriber(opt_keyexpr)
+                    .callback(callback)
+                    .allowed_origin(Locality::Remote)
+                    .res_async()
+                    .await
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    pub async fn publish(&self, key_expr: &str) -> ZResult<zenoh::publication::Publisher<'static>> {
+        debug!("Creating Publisher on {}", key_expr);
+
+        match self.make_keyexpr(key_expr).await {
+            Ok(opt_keyexpr) => {
+                self.session
+                    .declare_publisher(opt_keyexpr)
+                    .allowed_destination(Locality::Remote)
+                    .congestion_control(CongestionControl::Block)
+                    .res_async()
+                    .await
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    pub async fn make_queryable<Callback>(
+        &self,
+        key_expr: &str,
+        callback: Callback,
+    ) -> ZResult<zenoh::queryable::Queryable<'static, ()>>
+    where
+        Callback: Fn(zenoh::queryable::Query) + Send + Sync + 'static,
+    {
+        info!("Creating Queryable on {}", key_expr);
+
+        match self.make_keyexpr(key_expr).await {
```

Comment:
`make_keyexpr` is not necessary since the `declare_queryable` will already:
1) verify the correct format of the `keyexpr`;
2) automatically declare a keyexpr under the hood for optimization.